File: 10_21_2016_Binary-Tree-Preorder-Traversal/chang.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
 //divide and conquer
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Preorder in ArrayList which contains node values.
     */
    public ArrayList<Integer> preorderTraversal(TreeNode root) {
        // recursive dc
        ArrayList<Integer> result = new ArrayList<Integer>();
        
        if (root == null) {
            return result;
        }
        
        ArrayList<Integer> left = preorderTraversal(root.left);
        ArrayList<Integer> right = preorderTraversal(root.right);
        
        result.add(root.val);
        result.addAll(left);
        result.addAll(right);
        
        return result;
    }
}

File: 11_16_2016-convert-sorted-array-to-binary-search-tree/chang.py
class Solution:
    """
    @param A: a list of integer
    @return: a tree node
    """
    def sortedArrayToBST(self, A):
        if A is None or len(A) == 0:
            return None
        
        mid = len(A) / 2
        
        root = TreeNode(A[mid])
        
        root.left = self.sortedArrayToBST(A[0:mid])
        root.right = self.sortedArrayToBST(A[mid+1:len(A)])
        
        return root
